### ISSUE TITLE
Fixes to restart of IKernel  (Current the variable is undefined, and interrupt won't wait for  pending restart to finish) also Move initialization scripts for IW into Kernel or similar

### ISSUE THREAD

**Issue by DonJayamanne** [microsoft/vscode-jupyter#7985]
We've had a few issues with a test for `__file__` in interactive window

The problem was the cell was queued for execution even before IW completed running the initialization code.
In IKernel.ts we have initialization code, and that is guaranteed to run prior to any cell.

However IW listens to restart events of IKernel and runs the initailization code.
But then also sends code to the IW.
This is the sequence of events:
* Kernel restart started
* IW sends a cell to Kernel for execution
* Kernel restarted (event fired)
	* IW listens to this and starts initialization of __file__
* As soon as the restart is complete, kernel resumes execution of the cell sent to it
	* However the initialization of __file__ has not yet completed.

We need a way to tell IKernel.ts to wait for cell execution until some initialization has completed
Or we need IW to check if kernel was restared & completed or not, and accrodingly queue the messages (however all of this queuing of kernels is already done by IKernel today)

**Proposed Solution**
* Add notion of `post-restart-hooks`
* IW can provide a callback that needs to be called after a kernel is re-started (& then IKernel will know to wait on those hook for completion)

**Update - Above solution Doesn't work**
* The initialization code tries to run code against the kernel, but kernel is busy.
* Initialization should run using raw IKernelConnection without waiting on anythihng.
* Needs a better design.

**Comment by DonJayamanne** [comment]
@rchiodo @IanMatthewHuff /cc
